src/connector/src/sink/nats.rs

```diff
@@ -0,0 +1,172 @@
+use tokio_retry::Retry;
+
+use super::utils::chunk_to_json;
+use super::{DummySinkCommitCoordinator, SinkWriter, SinkWriterParam};
+use crate::common::NatsCommon;
+use crate::sink::{Result, Sink, SinkError, SINK_TYPE_APPEND_ONLY};
+
+pub const NATS_SINK: &str = "nats";
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize)]
+pub struct NatsConfig {
+    #[serde(flatten)]
+    pub common: NatsCommon,
+    // accept "append-only"
+    pub r#type: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct NatsSink {
+    pub config: NatsConfig,
+    schema: Schema,
+    is_append_only: bool,
+}
+
+// sink write
+pub struct NatsSinkWriter {
+    pub config: NatsConfig,
+    context: Context,
+    stream: Stream,
+    schema: Schema,
+}
+
+/// Basic data types for use with the nats interface
+impl NatsConfig {
+    pub fn from_hashmap(values: HashMap<String, String>) -> Result<Self> {
+        let config = serde_json::from_value::<NatsConfig>(serde_json::to_value(values).unwrap())
+            .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        if config.r#type != SINK_TYPE_APPEND_ONLY {
+            Err(SinkError::Config(anyhow!(
+                "Nats sink only support append-only mode"
+            )))
+        } else {
+            Ok(config)
+        }
+    }
+}
+
+impl NatsSink {
+    pub fn new(config: NatsConfig, schema: Schema, is_append_only: bool) -> Self {
+        Self {
+            config,
+            schema,
+            is_append_only,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for NatsSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type Writer = NatsSinkWriter;
+
+    async fn validate(&self, _client: Option<ConnectorClient>) -> Result<()> {
+        if !self.is_append_only {
+            return Err(SinkError::Nats("only support append-only mode".to_string()));
+        }
+        match self.config.common.build_context().await {
+            Ok(_jetstream) => {}
+            Err(error) => {
+                return Err(SinkError::Nats(format!(
+                    "validate nats sink error: {:?}",
+                    error
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    async fn new_writer(&self, _writer_env: SinkWriterParam) -> Result<Self::Writer> {
+        Ok(NatsSinkWriter::new(self.config.clone(), self.schema.clone()).await?)
+    }
+}
+
+impl NatsSinkWriter {
+    pub async fn new(config: NatsConfig, schema: Schema) -> Result<Self> {
+        Retry::spawn(
+            ExponentialBackoff::from_millis(100).map(jitter).take(3),
+            || async {
+                let context = config
+                    .common
+                    .build_context()
+                    .await
+                    .map_err(|e| SinkError::Nats(format!("nats sink error {}", e)))?;
+                let stream = config
+                    .common
+                    .build_or_get_stream(context.clone())
+                    .await
```

Comment:
Here when we create a new stream, I don't think we need to do the retry, since failing to create a nats sink writer is acceptable.